Mathlib/LinearAlgebra/ExteriorPower/Basic.lean

```diff
@@ -0,0 +1,427 @@
+
+@[inherit_doc]
+notation:100 "Λ[" R "]^" n:arg => ExteriorPower R n
+
+variable {M}
+
+namespace ExteriorPower
+
+/-! The canonical alternating from `Fin n → M` to `Λ[R]^n M`. -/
+
+/-- `ExteriorAlgebra.ιMulti` is the alternating map from `Fin n → M` to `Λ[r]^n M`
+induced by `ExteriorAlgebra.ιMulti`, i.e. sending a family of vectors `m : Fin n → M` to the
+product of its entries. -/
+def ιMulti : AlternatingMap R M ((Λ[R]^n) M) (Fin n) :=
+  AlternatingMap.codRestrict (ExteriorAlgebra.ιMulti R n) ((Λ[R]^n) M)
+  (fun _ => ExteriorAlgebra.ιMulti_range R n (by simp only [Set.mem_range, exists_apply_eq_apply]))
+
+@[simp] lemma ιMulti_apply (a : Fin n → M) :
+ιMulti R n a = ExteriorAlgebra.ιMulti R n a := by
+  unfold ιMulti
+  simp only [AlternatingMap.codRestrict_apply_coe]
+
+/-- The image of `ExteriorAlgebra.ιMulti R n` spans the `n`th exterior power. Variant of
+`ExteriorAlgebra.ιMulti_span_fixedDegree`, useful in rewrites. -/
+lemma ιMulti_span_fixedDegree :
+    Submodule.span R (Set.range (ExteriorAlgebra.ιMulti R n)) = (Λ[R]^n) M :=
+  ExteriorAlgebra.ιMulti_span_fixedDegree R n
+
+/-- The image of `ExteriorPower.ιMulti` spans `Λ[R]^n M`. -/
+lemma ιMulti_span :
+    Submodule.span R (Set.range (ιMulti R n)) = (⊤ : Submodule R ((Λ[R]^n) M)) := by
+  apply LinearMap.map_injective (Submodule.ker_subtype ((Λ[R]^n) M))
+  rw [LinearMap.map_span, ← Set.image_univ, Set.image_image]
+  simp only [Submodule.coeSubtype, ιMulti_apply, Set.image_univ, Submodule.map_top,
+    Submodule.range_subtype]
+  exact ExteriorAlgebra.ιMulti_span_fixedDegree R n
+
+/-- Two linear maps on `Λ[R]^n M` that agree on the image of `ExteriorPower.ιMulti`
+are equal. -/
+@[ext]
+lemma lhom_ext ⦃f : (Λ[R]^n) M →ₗ[R] N⦄ ⦃g : (Λ[R]^n) M →ₗ[R] N⦄
+    (heq : (LinearMap.compAlternatingMap f) (ιMulti R n) =
+    (LinearMap.compAlternatingMap g) (ιMulti R n)) : f = g := by
+  ext u
+  have hu : u ∈ (⊤ : Submodule R ((Λ[R]^n) M)) := Submodule.mem_top
+  rw [← ιMulti_span] at hu
+  apply Submodule.span_induction hu (p := fun u => f u = g u)
+  · exact fun _ h ↦ by
+      let ⟨a, ha⟩ := Set.mem_range.mpr h
+      apply_fun (fun F => F a) at heq; simp only [LinearMap.compAlternatingMap_apply] at heq
+      rw [← ha, heq]
+  · rw [LinearMap.map_zero, LinearMap.map_zero]
+  · exact fun _ _ h h' => by rw [LinearMap.map_add, LinearMap.map_add, h, h']
+  · exact fun _ _ h => by rw [LinearMap.map_smul, LinearMap.map_smul, h]
+
+/-! The universal property of the `n`th exterior power of `M`: linear maps from
+`Λ[R]^n M` to a module `N` are in linear equivalence with `n`-fold alternating maps from
+`M` to `N`. -/
+
+variable {R}
+
+/-- Auxiliary construction for the definition of `ExteriorPower.liftAlternating`. -/
+def liftAlternating_aux : (AlternatingMap R M N (Fin n)) →ₗ[R]
+    ((i : ℕ) → AlternatingMap R M N (Fin i)) :=
+  LinearMap.pi (fun i ↦ if h : i = n then by rw [h]; exact LinearMap.id else 0)
```

Comment:
I think it would be better to use `Pi.single` here, to avoid the `rw`.

Though looking again: is this just `LinearMap.single n`?

Reply:
Yes, it is just `LinearMap.single` ! Oops. I remove `liftAlternating_aux` since it just makes things more cumbersome.